[DOCS] Documents trained model auto-scaling

szabosteve opened the pull request:
## Overview

This PR:
- creates a new page about model autoscaling,
- provides tables for the autoscaling behavior in Cloud, (the content for Serverless will be part of another PR)
- gives instructions on how to optimize a model for search/ingest by using the API,
- adds link to ELSER, E5, and Deploy trained model pages that points to the autoscaling page,
- removes the corresponding ELSER limitation,
- reviews the text on [Deploy the model in your cluster](https://www.elastic.co/guide/en/machine-learning/master/ml-nlp-deploy-model.html) page,
 - updates the screenshot on the Deploy the model in your cluster page.


### Preview

* [Trained model autoscaling](https://stack-docs_bk_2795.docs-preview.app.elstc.co/guide/en/machine-learning/master/ml-nlp-auto-scale.html)

### Remarks

https://github.com/elastic/elasticsearch/pull/114038 removed a link to adaptive allocations docs from the Semantic text tutorial to avoid failing docs builds. The link needs to be updated and put back in the tutorial after this PR is merged.

arisonl commented:
This looks great @szabosteve, thanks for tagging me. A couple of questions:

- Agreed on a separate page for model scaling.
- Should we mention somewhere how to configure a deployment for search vs ingest via the API (1 thread for ingest, >=1 thread(s) for search)?
- Should we include a link to the pricing calculator (https://cloud.elastic.co/pricing) from the references you are making concerning managing cost?
- We need the equivalent documentation for serverless. 
- I am going to share with you a document with additional details that we may want to consider.

szabosteve commented:
@elasticmachine test this please

szabosteve commented:
@arisonl I amended the content based on your comment above.
* added a separate page for autoscale
* added a link that points to the pricing calculator
* gave instructions on how to configure for the ingest/search use case by using the API
* provided matrices for Cloud behavior

szabosteve commented on docs/en/stack/ml/nlp/ml-nlp-autoscaling.asciidoc:
```diff
@@ -0,0 +1,140 @@
+[cols="4*", options="header"]
+|==========
+| Level  | Allocations                                  | Threads | vCPUs
+| Low    | 0 to 2 dynamically                           | 1       | 0 to 2 dynamically 
```
@arisonl The docs say `1` but the latest decision was: "Zero for the Low usage level on both serverless and ESS." Please correct me if I'm wrong!

szabosteve commented on docs/en/stack/ml/nlp/ml-nlp-autoscaling.asciidoc:
```diff
@@ -0,0 +1,140 @@
+|==========
+| Level  | Allocations                                  | Threads                                            | vCPUs
+| Low    |  1                                           | 2                                                  | 2
+| Medium |  1 to 2 (if threads=16) dynamically          | maximum that the hardware allows (for example, 16) | 1 to 32 dynamically
```
`1 to 2 (if threads=16)`
Should we elaborate on this condition?

szabosteve commented:
run docs-build

szabosteve commented on docs/en/stack/ml/nlp/ml-nlp-autoscaling.asciidoc:
```diff
@@ -0,0 +1,139 @@
+
+[discrete]
+=== Deployments in Cloud optimized for search
+
```
For search, we try to maximize the number of threads.

szabosteve commented on docs/en/stack/ml/nlp/ml-nlp-autoscaling.asciidoc:
```diff
@@ -0,0 +1,139 @@
+
+[discrete]
+=== Deployments in Cloud optimized for ingest
+
```
For ingest, we try to maximize number of allocations.

szabosteve commented on docs/en/stack/ml/nlp/ml-nlp-autoscaling.asciidoc:
```diff
@@ -0,0 +1,145 @@
+In case of ingest-optimized deployments, we maximize the number of model allocations.
+
+
+
```
```suggestion
```

szabosteve commented on docs/en/stack/ml/nlp/ml-nlp-autoscaling.asciidoc:
```diff
@@ -0,0 +1,144 @@
+* the use case you optimize the model deployment (ingest or search)
+* whether adaptive resources are enabled or disabled (dynamic or static resources)
+
+If you use {es} on-premises, adaptive resources behavior is fully dynamic and highly dependent on the hardware configuration.
```
@darnautov Should we / can we provide any more information about this scenario?

darnautov replied:
`adaptive resources behavior is fully dynamic...` - I think we should rephrase this. vCPUs level ranges are derived based on the ml processors limits (https://www.elastic.co/guide/en/elasticsearch/reference/current/get-ml-info.html), not the adaptive resources behaviour.

jan-elastic commented on docs/en/stack/ml/nlp/ml-nlp-autoscaling.asciidoc:
```diff
@@ -0,0 +1,144 @@
+* trough APIs by enabling adaptive allocations
+* in {kib} by enabling adaptive resources
+
+IMPORTANT: You must enable {cloud}/ec-autoscaling.html[deployment autoscaling] to use model autoscaling.
```
This is not entirely true. You can just allocate one ML node and use adaptive allocations. Adaptive allocations will allocate more or less of the ML node depending on the load. However, if adaptive allocations decides it needs more than the available resources (the ML node), it cannot scale to that.

Adaptive allocations certainly works better in combination with ML autoscaling.

I'd make this a (strong) recommendation, but not say that you must enable it, since that's not true.

jan-elastic commented on docs/en/stack/ml/nlp/ml-nlp-autoscaling.asciidoc:
```diff
@@ -0,0 +1,144 @@
+The used resources for trained model deployments depends on three factors:
+
+* your cluster environment (Serverless, Cloud, or on-premises)
+* the use case you optimize the model deployment (ingest or search)
```
the use case you optimize the model deployment **for**

jan-elastic commented on docs/en/stack/ml/nlp/ml-nlp-autoscaling.asciidoc:
```diff
@@ -0,0 +1,144 @@
+| Level  | Allocations                                  | Threads | vCPUs
+| Low    | 0 to 2 dynamically                           | 1       | 0 to 2 dynamically 
+| Medium | 1 to 32 dynamically                          | 1       | 1 to the smaller of 32 or the limit set in the Cloud console, dynamically
+| High   | 1 to limit set by Cloud console, dynamically | 1       | 1 to limit set by Cloud console, dynamically
```
by -> in the (like prev line)

jan-elastic replied:
2x

jan-elastic replied:
also underneath

jan-elastic commented on docs/en/stack/ml/nlp/ml-nlp-autoscaling.asciidoc:
```diff
@@ -0,0 +1,144 @@
+|==========
+| Level  | Allocations                                              | Threads | vCPUs
+| Low    | 2 if available, otherwise 1, statically                  | 1       | 2 if available
+| Medium | 32 if available, otherwise maximum available, statically | 1       | 32 if available
```
Why say here "if available", instead of the more clear "the smaller of 32 or the limit set in the Cloud console" (like above)

jan-elastic commented on docs/en/stack/ml/nlp/ml-nlp-autoscaling.asciidoc:
```diff
@@ -0,0 +1,144 @@
+| Level  | Allocations                                  | Threads                                            | vCPUs
+| Low    |  1                                           | 2                                                  | 2
+| Medium |  1 to 2 (if threads=16) dynamically          | maximum that the hardware allows (for example, 16) | 1 to 32 dynamically
+| High   |  1 to limit set by Cloud console, dynamically| maximum that the hardware allows (for example, 16) | 1 to limit set by Cloud console, dynamically
```
To be precise, you don't set an allocations limit in the cloud console, but a vCPU limit. The leads to an allocations limit of vCPU limit divided by the number of threads. Good luck phrasing that nicely!

jan-elastic commented on docs/en/stack/ml/nlp/ml-nlp-autoscaling.asciidoc:
```diff
@@ -0,0 +1,144 @@
+[[nlp-model-adaptive-allocations]]
+== Enabling autoscaling through APIs - adaptive allocations
+
+Model allocations are independent units of work for NLP tasks.
```
Not sure about this sentence. I'd say model allocations can independently process units of work, not are units of work.

jan-elastic commented on docs/en/stack/ml/nlp/ml-nlp-autoscaling.asciidoc:
```diff
@@ -0,0 +1,144 @@
+== Enabling autoscaling through APIs - adaptive allocations
+
+Model allocations are independent units of work for NLP tasks.
+The numbers of threads and allocations you can set manually for a model remain constant even when not all the available resources are fully used or when the load on the model requires more resources.
```
I'd say: If you set the numbers of threads and allocations manually, they remain constant (...)".

jan-elastic commented on docs/en/stack/ml/nlp/ml-nlp-autoscaling.asciidoc:
```diff
@@ -0,0 +1,144 @@
+You can enable adaptive allocations by using:
+
+* the Create inference endpoint API for {ref}/infer-service-elser.html[ELSER], {ref}/infer-service-elasticsearch.html[E5 and models uploaded through Eland] that are used as {infer} services.
+* the {ref}/start-trained-model-deployment.html[start trained model deployment] or {ref}/update-trained-model-deployment.html[update trained model deployment] APIs for trained models that are deployed on {ml} nodes.
```
Nit: "Create inference endpoint API" is written with a capital, while "start trained model deployment API" is not. please unify

jan-elastic commented on docs/en/stack/ml/nlp/ml-nlp-autoscaling.asciidoc:
```diff
@@ -0,0 +1,144 @@
+[cols="4*", options="header"]
+|==========
+| Level  | Allocations                                  | Threads | vCPUs
+| Low    | 0 to 2 dynamically                           | 1       | 0 to 2 dynamically 
```
Almost everywhere "if available" is added, Why not here? It also applies to the small numbers.

jan-elastic commented on docs/en/stack/ml/nlp/ml-nlp-elser.asciidoc:
```diff
@@ -449,13 +452,17 @@ To achieve the best results, it's recommended to clean the input text before gen
-include::ml-nlp-shared.asciidoc[tag=ml-nlp-adaptive-allocations]
+* Use two ELSER {infer} endpoints: one optimized for ingest and one optimized for search.
+* If quick response time is important for your use case, keep {ml} resources available at all times by setting `min_allocations` to `1`.
+* Setting `min_allocations` to `0` can save on costs for non critical use cases or testing environments.
```
non-critical?

jan-elastic commented on docs/en/stack/ml/nlp/ml-nlp-autoscaling.asciidoc:
```diff
@@ -0,0 +1,144 @@
+[[ml-nlp-auto-scale]]
+= Trained model autoscaling
```
Who coined this term "trained model autoscaling". Haven't seen that one before.

I think I've lost track of how many names this features has...

jan-elastic commented on docs/en/stack/ml/nlp/ml-nlp-autoscaling.asciidoc:
```diff
@@ -0,0 +1,153 @@
+[[optimize-use-case]]
+=== Optimize for typical use cases by using adaptive allocations
+
+You can optimize your model deplyoment for typical use cases, such as search and ingest.
```
deplyoment

jan-elastic reviewed (approved):
lgtm

darnautov commented on docs/en/stack/ml/nlp/ml-nlp-deploy-models.asciidoc:
```diff
@@ -164,66 +164,34 @@ their deployment across your cluster under **{ml-app}** > *Model Management*.
+The resource usage levels behave differently depending on <<nlp-model-adaptive-resources, adaptive resources>> being enabled or disabled.
+When adaptive resources are disabled but {ml} autoscaling is enabled, vCPU usage of Cloud deployments derived from the Cloud console and functions as follows:
+
+* Low: This level limits resources to one vCPU, which may be suitable for development, testing, and demos depending on your parameters.
```
We set 2 vCPUs for low. 
```suggestion
* Low: This level limits resources to two vCPUs, which may be suitable for development, testing, and demos depending on your parameters.
```

darnautov commented on docs/en/stack/ml/nlp/ml-nlp-deploy-models.asciidoc:
```diff
@@ -164,66 +164,34 @@ their deployment across your cluster under **{ml-app}** > *Model Management*.
+
+* Low: This level limits resources to one vCPU, which may be suitable for development, testing, and demos depending on your parameters.
+It is not recommended for production use
+* Medium: This level limits resources to two vCPUs, which may be suitable for development, testing, and demos depending on your parameters.
```
32 for medium 

```suggestion
* Medium: This level limits resources to 32 vCPUs, which may be suitable for development, testing, and demos depending on your parameters.
```

darnautov commented on docs/en/stack/ml/nlp/ml-nlp-deploy-models.asciidoc:
```diff
@@ -164,66 +164,34 @@ their deployment across your cluster under **{ml-app}** > *Model Management*.
+It is not recommended for production use
+* Medium: This level limits resources to two vCPUs, which may be suitable for development, testing, and demos depending on your parameters.
+It is not recommended for production use.
+* High: This level may use the maximum number of vCPUs available for this deployment from the Cloud console.
```
As we do not have access to vCPUs ranges at the moment, current value for high static allocations is 128. cc @arisonl 

```suggestion
* High: This level may use the maximum number of vCPUs available for this deployment from the Cloud console.
```

darnautov commented on docs/en/stack/ml/nlp/ml-nlp-autoscaling.asciidoc:
```diff
@@ -0,0 +1,154 @@
+When you optimize for ingest, the throughput will be higher, which increases the number of {infer} requests that can be performed in parallel.
+When you optimize for search, the latency will be lower during search processes.
+
+* If you want to optimize for ingest, set the number of threads to `1` (`"num_threads": 1`).
```
this param is called differently in `trained_model` API, cc @jan-elastic 
```suggestion
* If you want to optimize for ingest, set the number of threads to `1` (`"thread_per_allocation": 1`).
```

darnautov commented on docs/en/stack/ml/nlp/ml-nlp-autoscaling.asciidoc:
```diff
@@ -0,0 +1,154 @@
+
+The used resources for trained model deployments depend on three factors:
+
+* your cluster environment (Serverless, Cloud, or on-premises)
```
Should we mention that for serverless Obs and Security projects adaptive resources are enabled at all times and we do not show the control in Kibana?

darnautov reviewed (approved):
LGTM
